Improve sharing of temporary_files.toml among multiple processes

evan-goode opened the pull request:
Related: https://github.com/rpm-software-management/dnf5/issues/2129.

This PR implements Petr's suggestion here:

https://github.com/rpm-software-management/dnf5/issues/1640#issuecomment-2934132836

> The original report highlights DNF's (mis)feature of removing "unneeded" downloaded packages from a cache after finishing a transaction. "Unneeded" now means all packages. This needs to change to "installed in the successful transaction". That means that downloaded, but never used packages will remain. Garbage-collecting them should be a job of an explicit "dnf clean packages" command.

After deleting the inbound RPMs, we also remove old entries in `temporary_files.toml` corresponding to files that no longer exist. This prevents `temporary_files.toml` from growing too large if the RPMs are removed by some other means, e.g. `dnf5 clean all`.

I added some locking for `temporary_files.toml`, keeping in mind that multiple processes might try to read/write it simultaneously.

pkratoch commented on libdnf5/base/transaction.cpp:
```diff
@@ -1258,24 +1258,46 @@ Transaction::TransactionRunResult Transaction::Impl::_run(
+
+            // Remove all nonexistent files from the temp files listing
+            std::vector<std::string> removed;
+            for (const auto & file : temp_files_set) {
```
What is the `temp_files_set` for? Why not iterate directly the `temp_files`?

evan-goode commented:
I changed it to remove `temporary_files.toml` when there are no longer any entries. Now that test is passing.

And I removed the locking of `temporary_files.toml`, I realized that it will be better to lock the entire cachedir instead of that single file. We can do that (and address the problem of two processes using the same package file) in a follow-up PR.

I dropped the commit to add the `keep_file` option for `Locker` but kept the new functions for blocking locks, I believe we will want those shortly.
